java/grpc-client/src/main/java/com/youtube/vitess/client/grpc/GrpcClientFactory.java

```diff
@@ -4,18 +4,257 @@
+    if (!keyStore.entryInstanceOf(alias, KeyStore.PrivateKeyEntry.class)) {
+      // There is no private key matching this alias
+      return null;
+    }
+
+    // Try loading the private key with the key password (which can be null)
+    try {
+      final char[] pass = keyPassword == null ? null : keyPassword.toCharArray();
+      final KeyStore.PrivateKeyEntry entry = (KeyStore.PrivateKeyEntry) keyStore.getEntry(alias, new KeyStore.PasswordProtection(pass));
+      return new PrivateKeyWrapper(entry.getPrivateKey(), keyPassword, entry.getCertificateChain());
+    } catch (UnrecoverableEntryException e) {
+        // Password invalid.  Try using the keystore password, and let the UnrecoverableEntryException bubble up
+        // this time if that password doesn't work either.
+        final char[] pass = keyStorePassword == null ? null : keyStorePassword.toCharArray();
+        final KeyStore.PrivateKeyEntry entry = (KeyStore.PrivateKeyEntry) keyStore.getEntry(alias, new KeyStore.PasswordProtection(pass));
+        return new PrivateKeyWrapper(entry.getPrivateKey(), keyPassword, entry.getCertificateChain());
+    }
+  }
+
+  /**
+   * TODO: Document
+   *
+   * @param keyStore
+   * @param keyStorePassword
+   * @param keyPassword
+   * @return
+   * @throws KeyStoreException
+   * @throws NoSuchAlgorithmException
+   */
+  private PrivateKeyWrapper loadPrivateKeyEntry(final KeyStore keyStore, final String keyStorePassword,
+                                                  final String keyPassword)
+          throws KeyStoreException, NoSuchAlgorithmException {
+    final Enumeration<String> aliases = keyStore.aliases();
+    while (aliases.hasMoreElements()) {
+      final String alias = aliases.nextElement();
+      try {
+        return loadPrivateKeyEntryForAlias(keyStore, alias, keyStorePassword, keyPassword);
+      } catch (UnrecoverableEntryException e) {
+          // Neither the key password nor keystore password could unlock this particular alias.  Not unexpected... just
+          // continue trying the other aliases.
+      }
+    }
+    return null;
+  }
+
+  /**
+   * TODO: Document
+   */
+  public static class TlsOptions {
+    private File keyStore;
+    private String keyStorePassword;
+    private String keyAlias;
+    private String keyPassword;
+    private File trustStore;
+    private String trustStorePassword;
+    private String trustAlias;
+
+    public TlsOptions keyStorePath(String keyStorePath) {
+      if (keyStorePath != null) {
+        this.keyStore = new File(keyStorePath);
+      }
+      return this;
+    }
+
+    public TlsOptions keyStorePassword(String keyStorePassword) {
+      this.keyStorePassword = keyStorePassword;
+      return this;
+    }
+
+    public TlsOptions keyAlias(String keyAlias) {
+      this.keyAlias = keyAlias;
+      return this;
+    }
+
+    public TlsOptions keyPassword(String keyStorePassword) {
+      this.keyPassword = keyPassword;
```

Comment:
This is supposed to be the `keyPassword` mutator, there's a separate one above for `keyStorePassword`.  There's just a misleading typo in the parameter name here, I've corrected it.